### ISSUE TITLE
`unconditional_recursion` lint doesn't work with struct update syntax

### ISSUE THREAD

**Issue by keiichiw** [rust-lang/rust#78474]
### Code

I tried this code which causes a stack overflow due to a recursive call of `Default::default()` I mistakenly added.

```rust
struct S {
    x: i32,
    y: i32,
}

fn f() -> i32 {
    42
}

impl Default for S {
    fn default() -> S {
        S {
            x: f(),
            ..Default::default() // cause stack overflow
        }
    }
}

fn main() {
    let _ = S::default();
}

```

I expected to see the `unconditional_recursion` warning. However, the compiler didn't complain it.

### Version it worked on

It most recently worked on: **1.43.0** showed the warning. (ref. [the compiler explorer](https://rust.godbolt.org/z/s9qnGP))

### Version with regression

The warning isn't shown on **1.44.0** or later, including the current stable 1.47.0.

**Comment by jonas-schievink** [comment]
I think this was caused by https://github.com/rust-lang/rust/pull/70822, but it looks like expected behavior: If `f` diverges, the self-recursion never happens.

**Comment by jonas-schievink** [comment]
Though maybe the lint should assume that functions returning an inhabited type won't diverge (since otherwise they should be marked `-> !`). I can definitely see why people would write code like this, so linting on it makes sense.

**Comment by lcnr** [comment]
I personally feel that we should lint for functions if they don't return an uninhabited type.
Even if someone wants to exit recursion by unwinding I do feel that a lint is still appropriate here
as it is really bad style imo.

**Comment by keiichiw** [comment]
Thanks for the quick response. Yeah, it'd be natural to expect a function whose return type is not `!` to return eventually.

Probably, we might want to have the following rules?:
1. Lint for self-recursion unless a function returning an inhabited type is called in advance.
2. If a non-diverge function is called before a self-recursion, show a message like "if this `f` diverges, please annotate it." in addition to the `unconditional_recursion` warning.

1 would help my case and 2 would help the case of #54444. (2 is not necessary, though.)

**Comment by spastorino** [comment]
Assigning `P-medium` as [discussed as part of the Prioritization Working Group procedure](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/I-prioritize.20.2378474.20.60unconditional_recursion.60.20lint.20doesn't.20.E2.80.A6) and removing `I-prioritize`.
